### PR TITLE
Some improvements to ArrayHelper - enforce the Utility pattern

### PR DESCRIPTION
**Pull request by eirslett**
They are two separate commits - accept both, cherry-pick one of them, or reject it all. ;-)

The first commit refactors ArrayHelper::sortObjects(), by using an anonymous function instead of a separate/protected method. This also removes the protected properties in the class - by looking at the code I guess they were there primarily to get around the limits of earlier PHP versions (no anonymous functions). The scope is reduced and the class is now stateless.

The second commit adds a final modifier to the class, to prevent extension. (I would say an "abstract" modifier _encourages_ extension, despite its intention). A private constructor enforces the Utility class pattern.

These commits should help protect ArrayHelper against "abuse".

(Sorry about the indentation typo in 63c9407, I'll fix and rebase if you want.)

### CODE REVIEW

File: vendor/Joomla/Utilities/ArrayHelper.php [hunk 1]
```diff
@@ -15,39 +15,12 @@
...
+final class ArrayHelper
 {
-	/**
-	 * Option to perform case-sensitive sorts.
-	 *
-	 * @var    mixed  Boolean or array of booleans.
-	 * @since  1.0
-	 */
-	protected static $sortCase;
+	// Private constructor to prevent instantiation of this class
```
Comment (florianv) [inline vendor/Joomla/Utilities/ArrayHelper.php]: It should be commented using the dockblock model we use in the other methods.

File: vendor/Joomla/Utilities/ArrayHelper.php [hunk 2]
```diff
@@ -15,39 +15,12 @@
...
+	private function __construct() {
```
Comment (florianv) [inline vendor/Joomla/Utilities/ArrayHelper.php]: The bracket should go on the next line.

**Comment by florianv** [comment]
Looks nice. I commented a few code style issues.

I think it would be nice to add the array typehint in all methods expecting an array, and remove the if is_array stuff.

For example `public static function toString($array = null...`, should be `public static function toString(array $array`.

**Comment by dongilbert** [comment]
I agree with that Florian. Much better approach.

**Comment by florianv** [comment]
I think some of these functions were created before PHP5, so it's nice to refresh them a bit :)

**Comment by eirslett** [comment]
There we go, type hinting added.
I also changed &$array to $array in some places, so arguments are passed by value, not by reference. (Since most methods in the class don't apply side effects to the argument, it doesn't make sense to pass them by reference) This will also make the code faster and use less memory.

The only exception is toInteger, which takes the argument as a reference, does its things on that reference and then returns void. I think that's a strange way to do it. Maybe it was the preferred way when this class was written? But it's no use changing it now as it's already being used "in the wild".

And, finally, a few lines were saved using array_map :-) (gotta love functional programming)

**Comment by florianv** [comment]
I think this function is odd :

`public static function toInteger(&$array, $default = null)`

The default value is here in case $array is not an array but that should never be the case (I mean these functions operate on arrays).

So I think it could become : 

`public static function toInteger(array &$array)`

**Comment by eirslett** [comment]
Is API backwards-compatibility an issue with all this stuff? Do we have to take into account 3rd party extensions that use them? Or will the Framework be a "fresh start"?

**Comment by florianv** [comment]
Well, I'm not 100% sure about the BC. I'm suggesting things, but it's better to wait a maintainer confirms before modifying your code.

**Comment by dongilbert** [comment]
The framework is the opportunity to make this stuff right. If you see oddities, then fix them with a PR. We are not worried about BC with this stuff right now.

So, I would make the `toInteger` method work the same as the others. Don't take a reference, and return the modified array.

**Comment by eirslett** [comment]
I can only find 1 actual usage of toInteger in the framework - it's in Joomla\Form\Form, and it doesn't use the $default parameter.
What about removing toInteger altogether and replacing its usage:

``` php
ArrayHelper::toInteger($value);
```

with this oneliner:

``` php
$value = array_map('intval', $value);
```

**Comment by florianv** [comment]
I think this function can be useful. So IMO, I will keep it.

**Comment by ianmacl** [comment]
How does each handle nested items?

**Comment by eirslett** [comment]
There we go, toInteger returns the array now. Is there a place where backwards-incompatible changes are documented?

**Comment by dongilbert** [comment]
| How does each handle nested items?

That was my question / concern as well. @eirslett is the end result the same with both methods?

**Comment by eirslett** [comment]
Sorry, I didn't understand the question - which "each"?

At least the unit tests pass, so I guess the end result is the same.

**Comment by dongilbert** [comment]
I mean how the function worked previously, to how it works presently. I
think that was Ian's question as well. I like what you changed it to better
than what it was previously, just need to make sure it functions the same.
If the UT's pass, I'm fine with it.

**Comment by eirslett** [comment]
When I changed toInteger's behavior, I changed the unit test as well (and the usage in Joomla\Form\Form). The other functions were changed without touching the tests.

**Comment by ianmacl** [comment]
I was talking about:
ArrayHelper::toInteger($value);

vs

$value = array_map('intval', $value);

Is there any value that our toInteger method provides?  Or should we be pointing people to array_map?

**Comment by dongilbert** [comment]
I see. I was thinking yes, we should just use that hoping we could get rid
of the Joomla\Utilities dependency from Form. But we are using ArrayHelper
3 other times in that specific file, so I just figured keep it as it is.
Also, since there's been talk about a refactor of the Form package, it
might be wasted effort to change it now.

**Comment by davidhurley** [comment]
I thought we were refactoring the Form package completely. If that happens then it might be too much work to change this for just a short while.

**Comment by eirslett** [comment]
There is no nesting logic in ArrayHelper::toInteger. It's just some fallover behavior, in case $array is not an array. (But why should that behavior belong inside toInteger anyways? In general, I think ArrayHelper has too much defensive programming. It's doing input validation that really should be done elsewhere.)

**Comment by florianv** [comment]
@davidhurley I think to refactor the Form package there are three steps:

Refactoring Filter so it's possible to plug new strategies (atm it's hardcoded and you need to override the class).
Create a Validator package that can validate constraints.

There are different types of constraints:
- single value constraint ie. the value is a valid email (basically all current form rules but untied to the form case).
- multiple value constraints : example : The password and password verification fields are equal.

Once these two packages are working as standalone, then Form can be rewritten.

The form should only be an API to validate and filter forms, the rendering should be removed or implemented as a widget system.

These packages can also be used to validate/filter command line inputs (so not tied to the form case).

**Comment by dongilbert** [comment]
This can't be cleanly merged. Update from master and push, then this should be good to merge, IMO.

**Comment by eirslett** [comment]
There we go!
5284f65 has a strange diff, but all I did was change 2 lines to fix toInteger's documentation.

**Comment by dongilbert** [comment]
The unit tests failed on the pull request, but I think that was because of a bad push to master that broke them, which you merged into your code. Some more work has been done, and this needs to be updated one more time. Sorry about that, but if you update, I'll get it merged as soon as I see it. :|

**Comment by dongilbert** [comment]
I want to add an array_search that does partial matchings against the values, as this is a feature missing from php core. So, I can send you a PR to add that, and fix up the UT's in the process.

**Comment by eirslett** [comment]
Ok, go ahead :-)

**Comment by dongilbert** [comment]
@eirslett I've sent a PR to you to get everything up to date and move this along, as well as adding in my imprved arraySearch method. If you accept and merge the PR on your end, we can then accept and merge this one.

**Comment by eirslett** [comment]
Merged :)

**Pull request merged**